src/server.rs

```diff
@@ -24,6 +30,38 @@ fn parse_and_analyze(code: &str) -> flux::semantic::nodes::Package {
     .unwrap()
 }
 
+/// Take a lsp::Range that contains a start and end lsp::Position, find the
+/// indexes of those points in the string, and replace that range with a new string.
+fn replace_string_in_range(
+    mut contents: String,
+    range: lsp::Range,
+    new: String,
+) -> String {
+    let mut string_range: (usize, usize) = (0, 0);
+    let lookup = line_col::LineColLookup::new(&contents);
+    for i in 0..contents.len() {
+        let linecol = lookup.get(i);
+        if string_range.0 == 0 {
+            if linecol.0 == (range.start.line as usize)
+                && linecol.1 == (range.start.character as usize)
+            {
+                string_range.0 = i;
+            }
+        } else if linecol.0 == (range.end.line as usize)
+            && linecol.1 == (range.end.character as usize)
+        {
+            string_range.1 = i + 1; // Range is not inclusive.
+            break;
+        }
+    }
+    if string_range.1 == 0 {
+        error!("range end not found after range start");
+        return contents;
+    }
+    contents.replace_range(string_range.0..string_range.1, &new);
+    contents
+}
```

Comment:
If I understand it right, this seems like it would be slow since the `for` loop iterates over each character in `contents`. Is there a quicker way to get the linear offset into `contents` from `(line,character)` values provided by the `range`?

Will contents be the complete contents of a (potentially very large) file?

Reply:
I think that concern is reasonable, and I share it. I couldn't find a better dependency, and I certainly didn't want to engineer this string processing myself in a "nice-to-have" sort of patch (because line-endings and character encodings, etc). There's another patch that brings in `line_col`, and I discovered `line_col` by looking at rust-analyzer and related tools. I was mostly just being lazy and trying to find something that would work quickly. I'm not sure I'm stoked on this solution _either_, which is why I called it out. The file _could_ be large, yes, though flux files, at the scale where this might be an issue, aren't really a thing.

Honestly, if you're super concerned about it, I may just remove the functionality for now, as it's new in this lsp server version, and completely unsupported in our current lsp server.

Reply:
Ok, this seems reasonable. The flux files in our repo are never bigger than a few kb. If we identify it as an issue we may want to come back to it.